### PR TITLE
Feature/lm 2679 implement promise support in chain methods

### PR DESCRIPTION
**Pull request by anton-locomote**

### CODE REVIEW

File: lib/ChainFind.js [hunk 1]
```diff
@@ -124,6 +150,16 @@ function ChainFind(Model, opts) {
...
     },
+    onlyAsync: function () {
+      var context = this;
+
+      if (arguments.length && Array.isArray(arguments[0])) {
+        opts.only = arguments[0];
+      } else {
+        opts.only = Array.prototype.slice.apply(arguments);
+      }
+      return new Bluebird.resolve(context);
```
Comment (tttmaximttt) [inline lib/ChainFind.js]: Bluebird better rename to `Promise`

File: lib/ChainFind.js [hunk 2]
```diff
@@ -135,17 +171,44 @@ function ChainFind(Model, opts) {
       this.only(_.difference(Object.keys(opts.properties), omit));
       return this;
     },
+    omitAsync: function () {
```
Comment (tttmaximttt) [inline lib/ChainFind.js]: omit is sync method no need promisification

File: lib/ChainFind.js [hunk 3]
```diff
@@ -135,17 +171,44 @@ function ChainFind(Model, opts) {
...
+      }
+      this.only(_.difference(Object.keys(opts.properties), omit));
+
+      return new Bluebird.resolve(context);
+    },
     limit: function (limit) {
       opts.limit = limit;
       return this;
     },
+    limitAsync: function (limit) {
```
Comment (tttmaximttt) [inline lib/ChainFind.js]: limit is sync method no need promisification

File: lib/ChainFind.js [hunk 4]
```diff
@@ -135,17 +171,44 @@ function ChainFind(Model, opts) {
...
+      var context = this;
+      opts.limit = limit;
+      return new Bluebird.resolve(context);
+    },
     skip: function (offset) {
       return this.offset(offset);
     },
+    skipAsync: function (offset) {
```
Comment (tttmaximttt) [inline lib/ChainFind.js]: skip is sync method no need promisification

File: lib/ChainFind.js [hunk 5]
```diff
@@ -135,17 +171,44 @@ function ChainFind(Model, opts) {
...
+      var skipOffset = this.offset(offset);
+      return new Bluebird.resolve(skipOffset);
+    },
     offset: function (offset) {
       opts.offset = offset;
       return this;
     },
+    offsetAsync: function (offset) {
```
Comment (tttmaximttt) [inline lib/ChainFind.js]: offset is sync method no need promisification

File: lib/ChainFind.js [hunk 6]
```diff
@@ -157,13 +220,35 @@ function ChainFind(Model, opts) {
       }
       return this;
     },
+    orderAsync: function (property, order) {
```
Comment (tttmaximttt) [inline lib/ChainFind.js]: order is sync method no need promisification

File: lib/ChainFind.js [hunk 7]
```diff
@@ -157,13 +220,35 @@ function ChainFind(Model, opts) {
...
+      return new Bluebird.resolve(context);
+    },
     orderRaw: function (str, args) {
       if (!Array.isArray(opts.order)) {
         opts.order = [];
       }
       opts.order.push([ str, args || [] ]);
       return this;
     },
+    orderRawAsync: function (str, args) {
```
Comment (tttmaximttt) [inline lib/ChainFind.js]: orderRaw is sync method no need promisification

File: lib/ChainFind.js (hunk 1 above)
Comment (anton-locomote) [reply lib/ChainFind.js]: Done in new commit https://github.com/locomote/node-orm2/pull/13/commits/e52a651f182b56e7b01da9d77f074071b07181b4

File: lib/ChainFind.js (hunk 2 above)
Comment (anton-locomote) [reply lib/ChainFind.js]: Removed method and tests in new commit e52a651

File: lib/ChainFind.js (hunk 3 above)
Comment (anton-locomote) [reply lib/ChainFind.js]: Removed method and tests in new commit e52a651

File: lib/ChainFind.js (hunk 4 above)
Comment (anton-locomote) [reply lib/ChainFind.js]: Removed method and tests in new commit e52a651

File: lib/ChainFind.js (hunk 5 above)
Comment (anton-locomote) [reply lib/ChainFind.js]: Removed method and tests in new commit e52a651

File: lib/ChainFind.js (hunk 6 above)
Comment (anton-locomote) [reply lib/ChainFind.js]: Removed method and tests in new commit e52a651

File: lib/ChainFind.js (hunk 7 above)
Comment (anton-locomote) [reply lib/ChainFind.js]: Removed method and tests in new commit e52a651

File: lib/ChainFind.js [hunk 8]
```diff
@@ -209,6 +247,44 @@ function ChainFind(Model, opts) {
...
+          conditions.or.push(or);
+        }
+          return resolve(opts.driver.remove(opts.table, conditions, function(err, result) {
+            if (err) return reject (err);
+            return resolve (result);
+          }));
+        });
+      });
+    },
+
```
Comment (dxg) [inline lib/ChainFind.js]: We've duplicated a lot of code in this PR. Can we make this simpler?

Comment (anton-locomote) [reply lib/ChainFind.js]: In find() method I'm mot sure that we can - because, it has strange logic even for callbacks.

**Comment by tttmaximttt** [comment]
In cases where code duplicates you should move duplicate code in function and use this function in both

**Comment by anton-locomote** [comment]
@tttmaximttt But where did you find code duplicates in last commit? :)

**Comment by tttmaximttt** [comment]
@anton-locomote not in last commit but in whole project in each function what not use promisified

**Comment by tttmaximttt** [comment]
this request newer pass code verification on original node-orm repo, the use code verification for detect duplicates of code

**Comment by anton-locomote** [comment]
Ok, understand

**Comment by anton-locomote** [comment]
@dxg @tttmaximttt Please, check new commit 11af0eb related to the code duplicates comments.

File: lib/ChainFind.js (hunk 8 above)
Comment (anton-locomote) [reply lib/ChainFind.js]: According to Max comment - I understood that main idea of this comment not related to promisify, but it is about code duplication in general. Please, check last changes related to those comments.

File: lib/ChainFind.js [hunk 9]
```diff
@@ -116,6 +115,36 @@ function ChainFind(Model, opts) {
...
+      var context = this;
+
+      if(!opts.conditions) {
+        opts.conditions = {};
+      }
+
+      return new Promise(function(resolve) {
+        if (typeof _.last(args) === "function") {
+          cb = args.pop();
+        }
```
Comment (dxg) [inline lib/ChainFind.js]: Why does `findAsync` accept a callback?

File: lib/ChainFind.js [hunk 10]
```diff
@@ -116,6 +115,36 @@ function ChainFind(Model, opts) {
...
+          }
+          opts.conditions.__sql.push(args);
+        }
+
+        if (cb) {
+          chainRun(cb);
+        }
+
+        return resolve(context);
+      });
```
Comment (dxg) [inline lib/ChainFind.js]: We are still duplicating logic here. We should reuse code where possible. New code should only handle nuances of promises/callbacks.

File: lib/ChainFind.js [hunk 11]
```diff
@@ -175,6 +204,21 @@ function ChainFind(Model, opts) {
...
+        merge  : opts.merge
+      };
+      return new Promise (function (resolve, reject) {
+        opts.driver.count(opts.table, conditionsPrepared, mergeOption, function (err, data) {
+          if (err || !data.length) {
+            return reject(err);
+          }
+          return resolve(data[0].c);
+        });
+      });
```
Comment (dxg) [inline lib/ChainFind.js]: Large parts of this code are identical to the code in `count`. This should not be the case; we should reuse code else this creates tech debt and potential bugs in future when someone updates one part of the code but not the other.

File: lib/ChainFind.js (hunk 8 above)
Comment (dxg) [reply lib/ChainFind.js]: This code is still duplicate.. we should not reimplement what's already implemented; we just need to handle nuances of callbacks vs promises.

File: lib/ChainFind.js [hunk 12]
```diff
@@ -219,28 +306,33 @@ function ChainFind(Model, opts) {
...
+        Promise.reject(err);
+      });
+    },
+    lastAsync: function () {
+      return this.runAsync().then(function (items) {
+        return Promise.resolve(items && items.length > 0 ? items[items.length - 1] : null);
+      }).catch(function (err) {
+        Promise.reject(err);
+      });
+    },
```
Comment (dxg) [inline lib/ChainFind.js]: Why can't this use `Promise.promisify` ?
We've got duplicate logic here.

**Pull request closed without merging**